File: typescript/vaid-pop/src/conformance.ts
/**
 * Packaged cross-language PoP conformance check — the firewall, shipped in the
 * tarball. TypeScript mirror of `vaid_pop.conformance` (Python) and the Rust
 * `tests/conformance.rs` / `tests/completion_conformance.rs` gates.
 *
 * A consumer who has only `npm install vaid-pop` can prove the primitive they
 * installed reproduces the frozen cross-language vectors byte-for-byte:
 *
 * ```console
 * $ npx vaid-pop-conformance      # exit 0 = PASS, 1 = BLOCKER
 * ```
 *
 * The vectors are the ones bundled with the package (`vectors/*.json`). The Rust
 * and Python gates assert the identical vectors; a repo-level drift check proves
 * every copy is byte-identical, so Rust output == Python output == TypeScript
 * output == vector.
 *
 * This module reads its vectors from disk and is therefore the one Node-only
 * entry point in the package; the library surface itself is runtime-neutral.
 */

import { readFileSync } from 'node:fs';

import {
  ed25519PublicKey,
  ed25519Sign,
  fromHex,
  toHex,
  type Ed25519Seed,
} from './crypto.js';
import { ASSURANCE_TIERS, type CompletionRecord } from './requestCompletion.js';
import type { RequestAuthPayload } from './requestAuth.js';
import { canonicalRequestSigningBytes, verifySignedPayload } from './vaidPop.js';

/** A cross-language byte-identity divergence — a hard BLOCKER. */
export class ConformanceError extends Error {
  override readonly name = 'ConformanceError';
}

/** The shape of a frozen vector file, as far as these checks read it. */
export interface Vector {
  digest_sha256_hex: string;
  ed25519: {
    private_key_seed_hex: string;
    public_key_hex: string;
    signature_hex: string;
  };
  input: Record<string, unknown>;
  assurance_tier_strings?: string[];
}

function loadVectorFile(name: string): Vector {
  // `../vectors` relative to this module resolves to the package's vectors/
  // directory from both `src/` (running from source) and `dist/` (installed).
  const url = new URL(`../vectors/${name}`, import.meta.url);
  return JSON.parse(readFileSync(url, 'utf8')) as Vector;
}

/** The operator-signing PoP vector bundled with the installed package. */
export function loadVector(): Vector {
  return loadVectorFile('operator_pop_v1.json');
}

/** The completion-record vector bundled with the installed package. */
export function loadCompletionVector(): Vector {
  return loadVectorFile('completion_v1.json');
}

function seedOf(v: Vector): Ed25519Seed {
  return fromHex(v.ed25519.private_key_seed_hex);
}

function assertHex(label: string, got: string, want: string): void {
  if (got !== want) {
    throw new ConformanceError(
      `${label} diverged from the frozen vector — BLOCKER\n  got    = ${got}\n  vector = ${want}`,
    );
  }
}

/** JCS + SHA-256 over the camelCase `RequestAuthPayload` == the frozen digest. */
export function checkDigest(v: Vector): void {
  const digest = canonicalRequestSigningBytes(v.input as unknown as RequestAuthPayload);
  assertHex('TypeScript PoP digest', toHex(digest), v.digest_sha256_hex);
  if (digest.length !== 32) {
    throw new ConformanceError(`digest must be 32 bytes, got ${digest.length}`);
  }
}

/**
 * The frozen seed derives the frozen public key, signs the frozen digest to the
 * frozen signature, and that signature verifies through the public verifier
 * path.
 */
export function checkSignature(v: Vector): void {
  const seed = seedOf(v);
  const publicKey = ed25519PublicKey(seed);
  assertHex('Ed25519 public key', toHex(publicKey), v.ed25519.public_key_hex);

  const digest = canonicalRequestSigningBytes(v.input as unknown as RequestAuthPayload);
  const signature = ed25519Sign(digest, seed);
  assertHex('TypeScript PoP signature', toHex(signature), v.ed25519.signature_hex);
  if (signature.length !== 64) {
    throw new ConformanceError(`signature must be 64 bytes, got ${signature.length}`);
  }

  // Round-trip through the verifier a third party would actually call.
  if (!verifySignedPayload(v.input, publicKey, fromHex(v.ed25519.signature_hex))) {
    throw new ConformanceError(
      'the frozen signature does not verify under verifySignedPayload — BLOCKER',
    );
  }
}

/**
 * The completion record: digest, signature, and — because this is the FIRST
 * vector with an enum — every `AssuranceTier` string, in order. The enum is the
 * most likely place for silent cross-language drift, since the string lands
 * inside the signed bytes.
 */
export function checkCompletion(v: Vector): void {
  const digest = canonicalRequestSigningBytes(v.input as unknown as CompletionRecord);
  assertHex('TypeScript completion digest', toHex(digest), v.digest_sha256_hex);

  const seed = seedOf(v);
  assertHex('completion public key', toHex(ed25519PublicKey(seed)), v.ed25519.public_key_hex);
  assertHex('TypeScript completion signature', toHex(ed25519Sign(digest, seed)), v.ed25519.signature_hex);

  const frozen = v.assurance_tier_strings;
  if (!frozen) throw new ConformanceError('completion vector is missing assurance_tier_strings');
  const ours = [...ASSURANCE_TIERS];
  if (JSON.stringify(ours) !== JSON.stringify(frozen)) {
    throw new ConformanceError(
      `AssuranceTier strings diverged from the frozen vector — BLOCKER\n` +
        `  got    = ${JSON.stringify(ours)}\n  vector = ${JSON.stringify(frozen)}`,
    );
  }
  if ((v.input as unknown as CompletionRecord).assuranceTier !== 'selfReported') {
    throw new ConformanceError(
      "the completion vector's declared tier must be the substantiated 'selfReported'",
    );
  }
}

/**
 * Run every firewall check against the bundled vectors. Throws
 * {@link ConformanceError} on any divergence; returns the vectors on PASS.
 */
export function run(): { pop: Vector; completion: Vector } {
  const pop = loadVector();
  checkDigest(pop);
  checkSignature(pop);
  const completion = loadCompletionVector();
  checkCompletion(completion);
  return { pop, completion };
}

/** CLI entry point: exit 0 on PASS, 1 on any divergence. */
export function main(): number {
  let result;
  try {
    result = run();
  } catch (error) {
    console.error(
      `CROSS-LANGUAGE PoP FIREWALL: MISMATCH — BLOCKER\n${(error as Error).message}`,
    );
    return 1;
  }
  console.log(
    'CROSS-LANGUAGE PoP FIREWALL: PASS — installed vaid-pop == frozen vectors, byte-for-byte\n' +
      `  request digest    = ${result.pop.digest_sha256_hex}\n` +
      `  request signature = ${result.pop.ed25519.signature_hex}\n` +
      `  completion digest = ${result.completion.digest_sha256_hex}`,
  );
  return 0;
}

File: crates/vaid-pop/src/bin/conformance.rs
//! Packaged cross-language PoP conformance check — the firewall, shipped in the crate.
//!
//! Rust counterpart of `vaid_pop.conformance` (Python) and
//! `vaid-pop/src/bin/conformance.ts` (TypeScript). Those two ship an executable so a
//! consumer with only `pip install vaid-pop` / `npm install vaid-pop` can prove the
//! primitive they installed reproduces the frozen vectors byte-for-byte. Rust had no
//! equivalent: its gates were `#[test]`s under `tests/`, and `cargo test` needs a
//! checkout, so the only Rust answer was "clone our repository and trust that it is
//! the source your crate was built from". That is a strictly weaker claim than the
//! other two languages make, and this binary closes it:
//!
//! ```console
//! $ cargo install vaid-pop            # builds the binary from the crate you fetched
//! $ vaid-pop-conformance              # exit 0 = PASS, 1 = BLOCKER
//! ```
//!
//! The vectors are `include_str!`d from `tests/vectors/`, which ships inside the
//! published `.crate` tarball. Because `cargo install` compiles from that tarball,
//! the bytes checked here are the bytes you received from crates.io — not a copy in
//! our repository, and not this comment. That is the whole point of a firewall: it
//! is only real if it runs against the artifact in your hands.
//!
//! WHAT IS AND IS NOT COVERED HERE. This crate is the signing *primitive*; the
//! request *signer* lives in `vaid-client`, which depends on this crate, so checking
//! the signer from here would invert the dependency. This binary therefore asserts
//! the primitive's contract — canonical digest, deterministic Ed25519 signature,
//! derived public key, and the completion-record vector including the `AssuranceTier`
//! enum-string drift guard. The signer path is asserted by
//! `cargo test -p vaid-client --test conformance`. Python's firewall covers both
//! because its signer ships in the same package; that is a packaging difference, not
//! a coverage gap in the standard.

use std::process::ExitCode;

use ring::signature::{Ed25519KeyPair, KeyPair};
use serde_json::Value;

use vaid_pop::request_auth::RequestAuthPayload;
use vaid_pop::request_completion::{AssuranceTier, CompletionRecord};
use vaid_pop::vaid_pop::canonical_request_signing_bytes;

// Vendored copies, byte-identical to every other language's (a CI `cmp` enforces
// that). They ship in the crate so this runs with no checkout and no network.
const OPERATOR_VECTOR_JSON: &str = include_str!("../../tests/vectors/operator_pop_v1.json");
const COMPLETION_VECTOR_JSON: &str = include_str!("../../tests/vectors/completion_v1.json");

/// A cross-language byte-identity divergence — a hard BLOCKER, never ship-anyway.
#[derive(Debug)]
struct ConformanceError(String);

impl std::fmt::Display for ConformanceError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

type Check = Result<(), ConformanceError>;

fn to_hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

fn unhex(s: &str) -> Vec<u8> {
    (0..s.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&s[i..i + 2], 16).expect("vector hex is well-formed"))
        .collect()
}

fn assert_hex(label: &str, got: &str, want: &str) -> Check {
    if got != want {
        return Err(ConformanceError(format!(
            "{label} diverged from the frozen vector — BLOCKER\n  got    = {got}\n  vector = {want}"
        )));
    }
    Ok(())
}

fn parse(label: &str, json: &str) -> Result<Value, ConformanceError> {
    serde_json::from_str(json)
        .map_err(|e| ConformanceError(format!("{label} vector does not parse — BLOCKER: {e}")))
}

fn seed_keypair(v: &Value) -> Result<Ed25519KeyPair, ConformanceError> {
    let seed = unhex(
        v["ed25519"]["private_key_seed_hex"]
            .as_str()
            .ok_or_else(|| ConformanceError("vector missing ed25519 seed — BLOCKER".into()))?,
    );
    Ed25519KeyPair::from_seed_unchecked(&seed)
        .map_err(|e| ConformanceError(format!("frozen seed is not a valid key — BLOCKER: {e}")))
}

/// JCS + SHA-256 over the REAL camelCase `RequestAuthPayload` reproduces the frozen
/// digest. Deserializing into the actual struct (rather than hashing the raw JSON)
/// is deliberate: it proves the field names and encodings this crate serializes are
/// the ones the vector froze, which is where cross-language drift actually happens.
fn check_operator_digest(v: &Value) -> Check {
    let payload: RequestAuthPayload = serde_json::from_value(v["input"].clone()).map_err(|e| {
        ConformanceError(format!(
            "vector input does not deserialize into a real RequestAuthPayload — BLOCKER: {e}"
        ))
    })?;
    let digest = canonical_request_signing_bytes(&payload);
    if digest.len() != 32 {
        return Err(ConformanceError(format!(
            "digest is {} bytes, expected 32 — BLOCKER",
            digest.len()
        )));
    }
    assert_hex(
        "operator canonical digest",
        &to_hex(&digest),
        v["digest_sha256_hex"].as_str().unwrap_or_default(),
    )
}

/// From the frozen seed, derive the same public key and the same deterministic
/// signature, then verify it. Ed25519 is deterministic (RFC 8032), so a divergence
/// here is a real contract break and never flakiness.
fn check_operator_signature(v: &Value) -> Check {
    let kp = seed_keypair(v)?;
    assert_hex(
        "operator public key",
        &to_hex(kp.public_key().as_ref()),
        v["ed25519"]["public_key_hex"].as_str().unwrap_or_default(),
    )?;
    let payload: RequestAuthPayload = serde_json::from_value(v["input"].clone())
        .map_err(|e| ConformanceError(format!("vector input invalid — BLOCKER: {e}")))?;
    let digest = canonical_request_signing_bytes(&payload);
    let sig = kp.sign(&digest);
    assert_hex(
        "operator signature",
        &to_hex(sig.as_ref()),
        v["ed25519"]["signature_hex"].as_str().unwrap_or_default(),
    )
}

/// The completion-record vector: same digest/signature discipline over a real
/// `CompletionRecord`.
fn check_completion(v: &Value) -> Check {
    let rec: CompletionRecord = serde_json::from_value(v["input"].clone()).map_err(|e| {
        ConformanceError(format!(
            "vector input does not deserialize into a real CompletionRecord — BLOCKER: {e}"
        ))
    })?;
    let digest = canonical_request_signing_bytes(&rec);
    assert_hex(
        "completion digest",
        &to_hex(&digest),
        v["digest_sha256_hex"].as_str().unwrap_or_default(),
    )?;
    let kp = seed_keypair(v)?;
    let sig = kp.sign(&digest);
    assert_hex(
        "completion signature",
        &to_hex(sig.as_ref()),
        v["ed25519"]["signature_hex"].as_str().unwrap_or_default(),
    )
}

/// THE ENUM DRIFT GUARD: every `AssuranceTier` must serialize to exactly the frozen
/// string, in order. Enum strings travel inside signed bytes, so a rename that looks
/// cosmetic in one language silently breaks byte-identity with the other two.
fn check_assurance_tiers(v: &Value) -> Check {
    let frozen: Vec<String> = v["assurance_tier_strings"]
        .as_array()
        .ok_or_else(|| ConformanceError("vector missing assurance_tier_strings — BLOCKER".into()))?
        .iter()
        .map(|s| s.as_str().unwrap_or_default().to_string())
        .collect();
    let serialized: Vec<String> = [
        AssuranceTier::SelfReported,
        AssuranceTier::CounterSigned,
        AssuranceTier::ThirdPartyAttested,
    ]
    .iter()
    .map(|t| {
        serde_json::to_value(t)
            .ok()
            .and_then(|x| x.as_str().map(str::to_string))
            .unwrap_or_default()
    })
    .collect();
    if serialized != frozen {
        return Err(ConformanceError(format!(
            "AssuranceTier strings diverged from the frozen vector — BLOCKER\n  got    = {serialized:?}\n  vector = {frozen:?}"
        )));
    }
    Ok(())
}

fn run() -> Result<(Value, Value), ConformanceError> {
    let operator = parse("operator_pop_v1", OPERATOR_VECTOR_JSON)?;
    let completion = parse("completion_v1", COMPLETION_VECTOR_JSON)?;
    check_operator_digest(&operator)?;
    check_operator_signature(&operator)?;
    check_completion(&completion)?;
    check_assurance_tiers(&completion)?;
    Ok((operator, completion))
}

fn main() -> ExitCode {
    match run() {
        Err(e) => {
            // Report on stderr so a mismatch is visible even when stdout is piped,
            // and exit non-zero so CI and `set -e` scripts stop.
            eprintln!("CROSS-LANGUAGE PoP FIREWALL: MISMATCH — BLOCKER\n{e}");
            ExitCode::FAILURE
        }
        Ok((o, c)) => {
            println!(
                "CROSS-LANGUAGE PoP FIREWALL: PASS — installed signer == frozen vectors, \
                 byte-for-byte\n  \
                 operator   digest    = {}\n  \
                 operator   signature = {}\n  \
                 completion digest    = {}\n  \
                 completion signature = {}",
                o["digest_sha256_hex"].as_str().unwrap_or_default(),
                o["ed25519"]["signature_hex"].as_str().unwrap_or_default(),
                c["digest_sha256_hex"].as_str().unwrap_or_default(),
                c["ed25519"]["signature_hex"].as_str().unwrap_or_default(),
            );
            ExitCode::SUCCESS
        }
    }
}
